Windows-Kompatibilität

Yasamato opened the pull request:
- yarn-Commands sind jetzt unter Windows ausführbar
- README Problemlösung für Windows angepasst
- App-Icon der Software-Challenge hinzugefügt (Nur für Windows verifiziert)

Yasamato committed:
Windows commands added

- Befehle für Windows hinzugefügt (unschön, wer Ideen?)
- bump rimraf
- Anleitung in der README für Windows-User hinzugefügt.

Yasamato committed:
App-Icon im Dock und Fenster anzeigen

Muss noch auf Linux & Mac getestet werden.

xeruf commented on README.md:
```diff
@@ -15,6 +15,13 @@
+### Für Windows-User
+
+- repository klonen
+- yarn und node installieren (evtl. muss auch noch [Kotlin](https://downlinko.com/download-install-kotlin-windows.html) sowie [Gradle](https://gradle.org/install/) manuell installiert werden)
```
warum sollte man kotlin installieren? Auch Gradle muss nicht installiert werden, das läuft über den Wrapper.

Yasamato replied:
Ich vermute dass das Windows-spezifisch liegt. Hat mich auch ziemlich verwundert, teste dass gleich noch mal

Yasamato replied:
Bei mir kommt es trotz Wrapper zur Fehlermeldung dass tools fehlen bei der gradle bzw. wenn das nachinstalliert wurde kotlin-installation. Ich vermute dass liegt irgendwo an meinem zerschossenem Windows-Setup...

Xerus2000 replied:
gib mir dazu bitte mal mehr details via Discord

xeruf commented on package.json:
```diff
@@ -8,11 +8,12 @@
     "start": "yarn install && yarn compile && (test \"hi\" && (test -d server || yarn update-server)) || echo I can\\'t test whether the server exists! && electron .",
+    "start-win": "yarn install && yarn compile && (test \"hi\" && (test -d server || yarn update-server-win)) || echo I can\\'t test whether the server exists! && electron .",
     "start-dev": "yarn install && yarn compile && electron . --dev",
-
```
diese newline war absichtlich da. Bitte passe deinen auto-formatter an.

xeruf commented on package.json:
```diff
@@ -8,11 +8,12 @@
+    "update-server-win": "git submodule update --remote && yarn compile-server-win",
     "compile-server": "cd socha && ./gradlew :server:deploy -q && cd .. && rimraf server && cp -r socha/server/build/runnable server && cp socha/player/build/libs/defaultplayer.jar .",
-
+    "compile-server-win": "cd socha && ./gradlew :server:deploy -q && cd .. && rimraf server && xcopy socha\\server\\build\\runnable server\\ /E && xcopy socha\\player\\build\\libs\\defaultplayer.jar .",
```
ich möchte eigentlich nicht verschiedene scripts für verschiedene OS haben. Gibt es echt kein `cp -r` in Windows? Dann branche bitte, genau wie im `start`-command auch. Das ist so geschrieben, weil es "test" in Windows nicht gibt.

Yasamato replied:
Hm... das wusste ich nicht, hatte im Internet nur die Tricks mit :; goto etc. gefunden. Werde das anpassen

xeruf commented on README.md:
```diff
@@ -15,6 +15,13 @@
 - `yarn update-server` ausführen, damit der server im Verzeichnis `server` verfügbar wird (dieser Schritt muss zum updaten des servers wiederholt werden)
 - GUI per `yarn start` starten
 
+### Für Windows-User
```
Ich sehe keinen Grund, einen eigenen Abschnitt zu erstellen, wenn die Punkte sich nur minimal unterscheiden, das macht das Readme unnötig lang. Stattdessen kannst du in den Punkten die OS-Besonderheiten anmerken, wenn es nach den anderen Änderungen überhaupt noch welche gibt.

Yasamato replied:
Sollte ich das in den Unterpunkt mögliche Probleme ganz unten einfach hinein schieben?

Yasamato committed:
windows-cmds mit test checken

Merged unique commands for windows into the original for linux.
adjusted README

SKoschnicke commented:
Wie waere es, statt einer Weiche fuer das Dateienkopieren eine JS Lib zu benutzen? https://www.npmjs.com/package/copy
So wie ich das sehe, muesste man das Paket nur in den Abhaengigkeiten referenzieren und koennte dann `copy socha/server/build/runnable server` auf allen Plattformen benutzen.

Genauso gibt es vielleicht etwas fuer `test`. Da gefaellt mir die Weiche auch ueberhaupt nicht.

SKoschnicke commented:
https://www.npmjs.com/package/does-exist

Agent77326 commented:
> Wie waere es, statt einer Weiche fuer das Dateienkopieren eine JS Lib zu benutzen? https://www.npmjs.com/package/copy
> So wie ich das sehe, muesste man das Paket nur in den Abhaengigkeiten referenzieren und koennte dann `copy socha/server/build/runnable server` auf allen Plattformen benutzen.
> 
> Genauso gibt es vielleicht etwas fuer `test`. Da gefaellt mir die Weiche auch ueberhaupt nicht.

Das Problem bei Copy hierbei ist, dass immernoch unterschieden werden muss zwischen \ und / bei Unix. / akzeptiert copy in Windows leider nicht.

SKoschnicke commented:
Vielleicht ist es besser, auch unter Windows eine richtige Shell zu benutzen. @Agent77326 wie benutzt du denn Git? Normalerweise ist da doch schon eine Shell bei. Laeuft es in der?

Agent77326 commented:
> Vielleicht ist es besser, auch unter Windows eine richtige Shell zu benutzen. @Agent77326 wie benutzt du denn Git? Normalerweise ist da doch schon eine Shell bei. Laeuft es in der?

Hm... ich war zwar noch nie ein Windows-Fan..... aber nach Experimenten mit der von git gelieferten bash-shell:
(getestet mit `yarn compile-server` -> `cd socha && ./gradlew :server:deploy -q && cd .. && rimraf server && copy socha/server/build/runnable server && copy socha/player/build/libs/defaultplayer.jar . `)

![image](https://user-images.githubusercontent.com/27964795/63767877-105fa580-c8cf-11e9-9a0d-ff74e8b837a1.png)
Funktioniert wunderbar, jedoch
![image](https://user-images.githubusercontent.com/27964795/63767910-279e9300-c8cf-11e9-9808-04428edb9755.png)
scheitert  yarn compile-server... also habe ich die Befehlskette zum debuggen zerlegt und bin auf das hier gestoßen:
![image](https://user-images.githubusercontent.com/27964795/63767966-4bfa6f80-c8cf-11e9-8bbc-8f1fb1b66245.png)

Also erneut versucht via `npm install rimraf` das zu beheben... jedoch wird dadurch rimraf nur in das npm: package-lock.json hinzugefügt, welche wir nicht benutzen. (Ausführung scheitert auch hier) Und ein manueller Eintrag in die package.json bei dependencies, schafft hier leider auch keine Abhilfe (ja, rimraf war seit anfangan bereits in devDependencies drinnen)..

In der Windows-CMD sieht es nicht anders aus:
![image](https://user-images.githubusercontent.com/27964795/63768353-481b1d00-c8d0-11e9-8933-0b172f0ef9f8.png)
natürlich scheitert auch hier `yarn compile-server`:
![image](https://user-images.githubusercontent.com/27964795/63768393-5d904700-c8d0-11e9-9370-63c714199e4f.png)
# Aber:
![image](https://user-images.githubusercontent.com/27964795/63768469-8c0e2200-c8d0-11e9-90e9-21cb132518b3.png)
???? Wenn ich die Pfade Windows-Backslash-typisch nehme mekert er nicht und es funktioniert auch `yarn start` mit Aufkommender GUI sowie Spiele können gespielt werden.

Selbige gilt natürlich auch für die git-shell:
![image](https://user-images.githubusercontent.com/27964795/63768621-fd4dd500-c8d0-11e9-963b-6867f63f33fd.png)

SKoschnicke commented:
Wenn du binaries von npm Paketen direkt in der Shell ausfuehren willst, muss entweder `node_modules/.bin` in deinem `PATH` sein oder du benutzt `./node_modules/.bin/<binary>` also z.B. `./node_modules/.bin/rimraf`. yarn und npm sollten aber den `PATH` richtig setzen, wenn die Scripte aus der package.json ausfuehren. Ich denke daher nicht, dass der "Syntaxfehler" durch das Fehlen von rimraf ausgeloest wird.

Agent77326 commented:
Ich habe meine Umgebungsvariablen gecheckt und sowohl yarn als auch node haben PATH-Variablen zu bin-ordner. Ich vermute dass das Problem irgendwo anders liegt, kann aber auch nicht sagen woran.

Durch das Herumexperimentieren habe ich letztlich nur herausgefunden, dass es mit den Backslashes im Pfad für Windows funktioniert, was aber eigentlich bereits durch die Ausführung des node-packages copy automatisch geschehen sollte, auch mit normalen Slash im Pfad..
![image](https://user-images.githubusercontent.com/27964795/63772834-376fa480-c8da-11e9-9b9f-2edbe7cab7e3.png)
Irgendwo hier muss ein Denkfehler sein, oder wir wissen einfach etwas wahrscheinlich ziemlich simples nicht...

SKoschnicke commented:
Vermutlich ist das `copy` package einfach nicht plattformunabhaengig und uebernimmt die Pfade so wie man ihm sie gibt. Wenn du die git bash benutzt, brauchen wir kein copy sondern koennen einfach cp -r benutzen.

Agent77326 commented:
Kann ich `"start": "yarn install && yarn compile && (test \"hi\" && (test -d server || yarn update-server)) || echo I can\\'t test whether the server exists! && electron .",` zu `"start": "yarn install && yarn compile && yarn update-server && electron .",` ändern? ich weiß nicht was genau an den test-commands das bei mir zerschießt, aber der Befehl kommt sonst nie bis zu `electron .`

SKoschnicke commented:
Klar, da der Rest eh nur in einer bash läuft, kann das test “hi” auch weg.

> Am 27.08.2019 um 15:17 schrieb Agent77326 <notifications@github.com>:
> 
> Kann ich "start": "yarn install && yarn compile && (test \"hi\" && (test -d server || yarn update-server)) || echo I can\\'t test whether the server exists! && electron .", zu "start": "yarn install && yarn compile && yarn update-server && electron .", ändern? ich weiß nicht was genau an den test-commands das bei mir zerschießt, aber der Befehl kommt sonst nie bis zu electron .
> 
> —
> You are receiving this because your review was requested.
> Reply to this email directly, view it on GitHub, or mute the thread.

Yasamato committed:
Removed branched winows/linux bash-command

Funktioniert jetzt nur noch in windows in einer bash-Umgebung!
Support for windows cmd and power-shell dropped.

SKoschnicke commented on package.json:
```diff
@@ -7,12 +7,12 @@
   "main": "main.js",
   "scripts": {
     "test": "yarn install && yarn compile",
-    "start": "yarn install && yarn compile && (test \"hi\" && (test -d server || yarn update-server)) || echo I can\\'t test whether the server exists! && electron .",
```
das `test -d server` soll bleiben, das ist das einzige sinnvolle `test`-command. Das testet naemlich, ob das Verzeichnis `server` existiert und fuehrt nur `yarn update-server` aus, wenn es noch nicht existiert.

xeruf replied:
das davor testet, ob das test-command überhaupt existiert, würde mich wundern wenn wir das einfach so entfernen können?

SKoschnicke replied:
Wir haben erstmal bash als requirement. Sonst haben wir das `cp` Problem. Da bringt es uns auch nix, wenn man auf das Vorhandensein von `test` testet.

Yasamato replied:
Der test, ob der test-command vorhanden ist führt zu einem Abbruch bei mir unter Windows, hat vielleicht mit dem Parameter zu tun, welcher nach dem nicht gefundenen command als irgendwas anderes behandelt wird, aber nicht als zugehöriger Parameter? Weil es ein Double-quoted String ist? K.a.....

Yasamato committed:
test for server in package.json

test -d server wieder hinzugefügt
(Note: in windows it will always execute yarn update-server as test is not available)

xeruf commented on README.md:
```diff
@@ -15,6 +15,9 @@
 - `yarn update-server` ausführen, damit der server im Verzeichnis `server` verfügbar wird (dieser Schritt muss zum updaten des servers wiederholt werden)
 - GUI per `yarn start` starten
 
+### Bei Problemen mit Windows-User
```
```suggestion
### Bei Problemen unter Windows
```

xeruf commented on README.md:
```diff
@@ -15,6 +15,9 @@
 - GUI per `yarn start` starten
 
+### Bei Problemen mit Windows-User
+Es kann eine manuelle Installation von [Kotlin](https://downlinko.com/download-install-kotlin-windows.html) sowie [Gradle](https://gradle.org/install/) erforderlich sein, sollte die Fehlermeldung von `missing tools` auftreten.
```
gib mir zu deinem Problem bitte mal mehr details via Discord

Yasamato replied:
Ich kann es leider nicht mehr reproduzieren... muss irgendwas mit meinem Windows-Setup gewesen sein...

Yasamato replied:
Ich würde den Teil der Readme einfach entfernen

Yasamato replied:
Ich habe es nach unten zu den möglichen Problemen verschoben sowie erweitert siehe https://github.com/CAU-Kiel-Tech-Inf/socha-gui/pull/26/commits/2e087718e64687a2407ad455c3874b62443add86

xeruf commented on package.json:
```diff
@@ -7,7 +7,7 @@
   "scripts": {
     "test": "yarn install && yarn compile",
-    "start": "yarn install && yarn compile && (test \"hi\" && (test -d server || yarn update-server)) || echo I can\\'t test whether the server exists! && electron .",
+    "start": "(test -d server || yarn update-server) && yarn install && yarn compile && electron .",
```
warum eine andere reihenfolge?

xeruf replied:
damit rimraf im update-server funktioniert, muss nämlich erst install aufgerufen werden

Yasamato replied:
Oh, ja mein Versehen... funktioniert btw trotzdem unter Windows

Agent77326 commented:
@Xerus2000 Ich weiß immer noch nicht was mein Problem mit dem Gradle-Wrapper war. Hab es gerade erneut getestet ohne die Installationen und jetzt kamen keine Fehlermeldungen mehr...

Abgesehen davon dass yarn meckert, dass rimraf nicht installiert ist, was mit `npm install rimraf` gefixt werden kann (bump to v2.7) siehe #25

Yasamato committed:
Fixed wrong order in package.json

Yasamato committed:
Windows-Problemlösungsansätze verschoben

Inhalt erweitert, sowie Erwähnung der fehlenden CMD/PowerShell Unterstützung.

Agent77326 commented:
Ich würde das in master und hive pullen damit ich auch hive bei mir kompillieren lassen kann.
